File: C#/Inheritance/Exercism - Wizards and Warriors.cs
// Link to the exercise: https://exercism.org/tracks/csharp/exercises/wizards-and-warriors

using System;

abstract class Character
{
    private readonly string _characterType;
    public bool vulnerability;
    protected Character(string characterType, bool vuln)
    {
        _characterType = characterType;
        vulnerability = vuln;
    }

    public abstract int DamagePoints(Character target);

    public virtual bool Vulnerable() =>
        false;

    public override string ToString() =>
        $"Character is a {_characterType}";
}

class Warrior : Character
{
    public Warrior() : base("Warrior", false)
    {
    }

    public override int DamagePoints(Character target) =>
        target.vulnerability ? 10 : 6;

    public override string ToString() =>
        base.ToString();

    public override bool Vulnerable() =>
        base.Vulnerable();
 
}

class Wizard : Character
{
    public Wizard() : base("Wizard", true)
    {
    }

    public override int DamagePoints(Character target) => 
        vulnerability ? 3 : 12;

    public void PrepareSpell() =>
        vulnerability = false;

    public override string ToString() =>
            base.ToString();

    public override bool Vulnerable() =>
        vulnerability;
}

File: Java/Inheritance/Exercism - Wizards and Warriors.java
// Link to the exercise: https://exercism.org/tracks/java/exercises/wizards-and-warriors

abstract class Fighter {
    private final String FIGHTER_TYPE = this.getClass().getName();
    protected boolean vulnerability = true;

    protected boolean isVulnerable() {
        return false;
    }

    public String toString() {
        return String.format("Fighter is a %s", FIGHTER_TYPE);
    }

    abstract int damagePoints(Fighter fighter);

}

class Warrior extends Fighter {
    protected final int MIN_DAMAGE = 6;
    protected final int MAX_DAMAGE = 10;

    @Override
    int damagePoints(Fighter fighter) {
        return fighter.vulnerability 
            ? MAX_DAMAGE
            : MIN_DAMAGE;
    }
}

class Wizard extends Fighter {
    protected final int MIN_DAMAGE = 3;
    protected final int MAX_DAMAGE = 12;

    @Override
    protected boolean isVulnerable() {
        return vulnerability;
    }

    @Override
    int damagePoints(Fighter fighter) {
        return vulnerability
            ? MIN_DAMAGE
            : MAX_DAMAGE;
    }

    void prepareSpell() {
        vulnerability = false;
    }
}
